### ISSUE TITLE
Closing/destroying sources when done

### ISSUE THREAD

**Issue by rubensworks** [RubenVerborgh/AsyncIterator#6]
In https://github.com/comunica/comunica/issues/144, we discussed adding a `closeSourceWhenDone` (_defaults to `true`_) to `TransformIterator `. This would make the transformed iterator call `close()` on its source when the transformed iterator has ended.

While this would work in practise, it may not be the cleanest solution, as `Readable#close` is not part of the [official interface](https://nodejs.org/api/stream.html#stream_readable_destroy_error).
Instead, [`Readable#destroy`](https://nodejs.org/api/stream.html#stream_readable_destroy_error) _is_ part of the interface, and can be used to achieve the same thing (and has been [available for a long time](https://github.com/nodejs/node-v0.x-archive/blob/cfcb1de130867197cbc9c6012b7e84e08e53d032/lib/fs.js#L1587-L1594)).

So I'm wondering if it would be better to instead use a property `destroySourceWhenDone` that calls `destroy()` on the source. This would probably require `AsyncIterator` to implement the `destroy()` method as well, which could simply be a proxy for `close()`.

**Comment by RubenVerborgh** [comment]
It seems that `destroy` has stronger semantics than `close`.

### destroy

> Destroy the stream, and emit 'error' and 'close'. After this call, the readable stream will release any internal resources and subsequent calls to push() will be ignored. Implementors should not override this method, but instead implement readable._destroy().

### close

> Stops the iterator from generating new items.
Already generated items or terminating items can still be emitted. After this, the iterator will end asynchronously.

Our options:
- implement an actual `destroy` on `AsyncIterator` (no real necessity, and can have some complexities)
- assume the existence of `close`
- try `close`
- try `close`, fall back to `destroy`

**Comment by rubensworks** [comment]
> implement an actual destroy on AsyncIterator (no real necessity, and can have some complexities)

I think this first option would be the most robust, and it strictly uses defined methods.

> try close, fall back to destroy

This last one might also be a good possibility, but I'm afraid that this may result in unexpected situations depending on the type of source streams.

In any case, it's your library, so it's your call in any case 😉

**Comment by RubenVerborgh** [comment]
>> implement an actual destroy on AsyncIterator (no real necessity, and can have some complexities)
> 
> I think this first option would be the most robust, and it strictly uses defined methods.

Yes, but adding a `destroy` method to this library with all the guarantees that come with it, might be complicated. I don't think it's easy to guarantee in general that destroyed iterators do not emit anything anymore, without adding checks that would slow down things. Technically speaking, we'd have to jump from `OPEN` to `ENDED` state, skipping `CLOSING` and `CLOSED`.

> In any case, it's your library, so it's your call in any case

Well, you'll be the one implementing it, so you get a say too :smile:
If you can implement `destroy` with the correct guarantees and without much impact, feel free to go ahead.

**Comment by rubensworks** [comment]
To summarize our offline talk:

Node documentation on `destroy`:
> After this call, the readable stream will release any internal resources and subsequent calls to push() will be ignored.

This makes it seem like `destroy` can do the cleanup _asynchronously_, which makes it possible to just implemented it like a `close` that will not emit any more elements, and will therefore not emit the `end` event anymore.

## Properties of all states

The following tables summarizes all current asynciterator states in terms of the following properties:

1. _Generate items_: Items can be prepared in an internal buffer (can be async).
2. _Emit items_: Items can be emitted using the `data` event or the `read()` method.
3. _Schedule new generations_: Scheduling of asynchronous item generation.

| State          | Generate items  | Emit items | Schedule new generations |
| ----------- | ---------------- | ----------| ----------------------------|
| `INIT`         | ✓                         | ✓              | ✓                                             |
| `OPEN`      | ✓                         | ✓              | ✓                                             |
| `CLOSING` | ✓                        | ✓              |                                                |
| `CLOSED`  |                            | ✓              |                                                |
| `ENDED`    |                            |                  |                                                 |

_(Feel free to edit this table directly)_

## Meaning of `destroy()` in this context

So, in terms of the discussed properties, calling a `destroy` should halt the generation of items (1), block any emission of items (2), and not allow any new item generations to be scheduled (3). In that sense, _destroying_ an asynciterator should immediately jump to the `ENDED` state (but not emit the `end` event).

## Difference with `close()`

As shown below, `close()` and `destroy()` are equal in terms of target states of an asynciterator. The only remaining difference is `close()` will set the state to `CLOSING`, allowing any remaining generation of items or buffered items to be gracefully finish. `destroy()` on the other hand, should also block the generation and emission of items, and therefore _skip_ the `CLOSING` and `CLOSED` states.

## Proposal for implementing `destroy()`

_Does nothing if the asynciterator was *ended* or *destroyed* already_

Introduction of a new `DESTROYED` state (parallel to `ENDED`), that has the same properties as `ENDED`, but will **not** emit an `end` event anymore. When in this state, `iterator.ended` will be false, but `iterator.destroyed` will be true.

Calling `destroy([error])`:

1. Set the state to `DESTROYED`.
2. Unregister listeners.
3. Clear any internal buffers (possibly asynchronously).
4. Emit an error event if an error object was passed.

I suggest changing the implementation of `_push` as follows:
```
BufferedIterator.prototype._push = function (item) {
  if (!this.destroyed) {
    if (this.ended)
      throw new Error('Cannot push after the iterator was ended.');
    this._pushedCount++;
    this._buffer.push(item);
    this.readable = true;
  }
};
```

(I assume some other places will also need to be changed to check both `this.ended` and `this.destroyed`)

## Alternative implementation

Same as above, but `this.ended` also takes into account the `DESTROYED` state. This would simplify several checks. The `_push` method would for example require no changes.

But this might be too confusing, as the iterator was not really _ended_ (no `end` event was emitted). So I would go for my first proposal.

**Comment by RubenVerborgh** [comment]
Currently, iterators always eventually are `ended`. Your proposal introduces a separate `destroyed` state that invalidates that assumption.

I think that makes sense, given that a destruction is not a proper ending.

Maybe it makes sense to have `DESTROYED = 5` (given that `ENDED = 4`), and introduce a `.done` property with `_state >= ENDED`.

Then the new situation is that iterators always end up `done`, being either `ended` or `destroyed`.

I wonder if there is a possibility of simplifying the check inside of `_push` by just inspecting `_state` directly. Only issue is then that deriving classes cannot override `destroyed` or `ended` (but maybe they shouldn't). We should just be careful about the performance of `_push`. Another possibility is to change the contract of `_push` such that pushing in `done` state silently fails, if that would be desired.

**Comment by rubensworks** [comment]
> Maybe it makes sense to have DESTROYED = 5 (given that ENDED = 4), and introduce a .done property with _state >= ENDED.

👍 

> Another possibility is to change the contract of _push such that pushing in done state silently fails, if that would be desired.

I think that would make sense. In most cases, silently failing a push is acceptable behaviour I guess. If users still want an explicit check, they'd have to do it themselves.

**Comment by RubenVerborgh** [comment]
Yeah, and that function not throwing will probably be better for performance too. Will be a major version change then though.

So how about we just surround everything with a `if (!this.done)` then?

In any case, green light to proceed with a PR in this way.

**Comment by rubensworks** [comment]
> So how about we just surround everything with a if (!this.done) then?

Yep, that should work!

> In any case, green light to proceed with a PR in this way.

Will continue working on it tomorrow :-)

**Comment by RubenVerborgh** [comment]
Following up in #7.

**Comment by rubensworks** [comment]
Maybe this shouldn't be closed just yet. #7 was just a part of this, the auto-closing has not been implemented yet, #7 was a requirement for that.

**Comment by RubenVerborgh** [comment]
Right.

**Comment by rubensworks** [comment]
Implemented in #8
